### PR TITLE
feat/limit-start-date

### PR DESCRIPTION
**Pull request by fredriklagerblad**
# Jira Issue: [IPC-227] 

## What?
- Start validating quotes's startDate
- Add a general guidelines that is run for all quotes
- Add startDate validation as a general guideline, not allowing it to be more than 2 years ahead
- To cover the Rapio case, where the startData is only sent in the sign request (where the guidelines aren't run), we add an excplitit validation there
- For rapio cases, we also validate that the startDate is not in the past  

## Why?
- Don't allow startDates to far in the future
- For external quotes (Rapio), we don't won't to allow startDates in the past

## Optional checklist
- [ ] Codescouted
- [x] Unit tests written
- [ ] Tested locally

[IPC-227]: https://hedvig.atlassian.net/browse/IPC-227

### CODE REVIEW

File: src/main/kotlin/com/hedvig/underwriter/service/SignServiceImpl.kt [hunk 1]
```diff
@@ -317,6 +318,9 @@ class SignServiceImpl(
                 if (startDate == null && currentInsurer == null) {
                     throw ErrorException(ErrorCodes.INVALID_STATE, "currentInsurer is required when startDate is null")
                 }
+                if (!it.isValidStartDate()) {
+                    throw ErrorException(ErrorCodes.INVALID_STATE, "Invalid startDate, too old or too far ahead in the future")
```
Comment (Michael-Duivestein) [inline src/main/kotlin/com/hedvig/underwriter/service/SignServiceImpl.kt:322]: It would appear that this exception will be thrown under three different conditions:

1. No start date
2. Start date is more than 2 years into the future
3. Start date is in the past.

Can we maybe change the error description so that it represents more than one of the three conditions?

Comment (fredriklagerblad) [reply src/main/kotlin/com/hedvig/underwriter/service/SignServiceImpl.kt:322]: No startDate should not trigger this exception, i.e. not considered invalid by the isValidStartDate() method.
(But no startDate and no currentInsurer would trigger, as it was before)

Comment (fredriklagerblad) [reply src/main/kotlin/com/hedvig/underwriter/service/SignServiceImpl.kt:322]: And the message covers the 2 other cases.

Comment (Michael-Duivestein) [reply src/main/kotlin/com/hedvig/underwriter/service/SignServiceImpl.kt:322]: Oops, didn't read the code properly. Sorry!

File: src/main/kotlin/com/hedvig/underwriter/service/guidelines/GeneralGuidelines.kt [hunk 2]
```diff
@@ -0,0 +1,27 @@
+package com.hedvig.underwriter.service.guidelines
+
+import com.hedvig.underwriter.model.Quote
+import com.hedvig.underwriter.model.QuoteData
+import com.hedvig.underwriter.model.isValidStartDate
+import com.hedvig.underwriter.service.guidelines.BreachedGuidelinesCodes.INVALID_STARTDATE
+import com.hedvig.underwriter.service.guidelines.BreachedGuidelinesCodes.OK
+
+// These validate general things, not directly connected to a specific type
+object GeneralGuidelines {
```
Comment (meletios) [inline src/main/kotlin/com/hedvig/underwriter/service/guidelines/GeneralGuidelines.kt:10]: Sweet

File: src/main/kotlin/com/hedvig/underwriter/serviceIntegration/memberService/dtos/DebtDto.kt [hunk 3]
```diff
@@ -1,6 +1,6 @@
 package com.hedvig.underwriter.serviceIntegration.memberService.dtos
 
-class DebtDto(
+data class DebtDto(
```
Comment (meletios) [inline src/main/kotlin/com/hedvig/underwriter/serviceIntegration/memberService/dtos/DebtDto.kt:3]: Is this related to this PR? 😅

**Review by meletios — APPROVED** [review]
Sweet, i have a naming suggestion

File: src/main/kotlin/com/hedvig/underwriter/model/Quote.kt [hunk 4]
```diff
@@ -80,6 +80,25 @@ val Quote.danishTravel
 val Quote.validTo
     get() = this.createdAt.plusSeconds(this.validity)!!
 
+fun Quote.isValidStartDate(): Boolean {
```
[the remark this answers is not in the document]
Comment (fredriklagerblad) [reply src/main/kotlin/com/hedvig/underwriter/model/Quote.kt:83]: 👍 I fix.

File: src/main/kotlin/com/hedvig/underwriter/serviceIntegration/memberService/dtos/DebtDto.kt (hunk 3 above)
Comment (fredriklagerblad) [reply src/main/kotlin/com/hedvig/underwriter/serviceIntegration/memberService/dtos/DebtDto.kt:3]: Sort of yes :-) 
Discovered when testing that the toString method of this class wasn't informational, since ot wasn't a data class ( like it should be)
